File: java/L800/LeetCode0877.java
package L800;

public class LeetCode0877 {
    public boolean stoneGame(int[] piles) {
        final int N = piles.length;
        int[][] dp = new int[N][N];
        int sum = 0;
        for (int pile : piles) {
            sum += pile;
        }
        int x = stones(piles, 0, N - 1, dp);
        return x > sum - x;
    }

    private int stones(int[] piles, int left, int right, int[][] dp) {
        if (left > right) {
            return 0;
        } else if (left == right) {
            return piles[left];
        } else if (dp[left][right] != 0) {
            return dp[left][right];
        }
        dp[left][right] = Math.max(stones(piles, left + 1, right, dp) + piles[left], 
                                   stones(piles, left, right - 1, dp) + piles[right]);
        return dp[left][right];
    }
}

File: cpp/L800/LeetCode0877.cc
#include <vector>
using std::vector;

class Solution {
public:
    bool stoneGame(vector<int>& piles) {
        int N = piles.size();
        vector<vector<int>> dp(N, vector<int>(N, 0));
        for (int i = 0;i < N;++i) {
            dp[i][i] = piles[i];
        }
        for (int i = N - 1;i >= 0;--i) {
            for (int j = i;j < N;++j) {
                int x1 = piles[i] - (i + 1 >= N ? 0 : dp[i + 1][j]);
                int x2 = piles[j] - (j - 1 < 0 ? 0 : dp[i][j - 1]);
                dp[i][j] = std::max(x1, x2);
            }
        }
        return dp[0][N - 1] > 0;
    }
};
